### ISSUE TITLE
Streaming Audio Example

### ISSUE THREAD

**Issue by nahap** [sonos/tract#114]
Thanks for your greate work!
I would love to see an audio streaming inference Example here.

TIA
Andy

**Comment by ronny-rentner** [comment]
I'd also like to see such an example :)

**Comment by kali** [comment]
There's hope! I can not share our in-house models, but I've been submitted a quartznet example. It causes a few difficulties, but I'm slowly getting them fixed.

**Comment by stexa** [comment]
Yeah and it would be nice to have some infos if running the model in the current examples with tvec! is allocating / realtime safe.

**Comment by kali** [comment]
tract is not real-time safe. I am not even sure how to define this term, actually. Most of the time, common neural networks are "well-behaved", performing the same operations from one run to the next. But there is no warranties.

tract performs allocation all the time, everywhere. Some attention has been given to remove some allocations from critically tights loops, but that's about it. From our experience, on our use-cases, the modern allocators (i.e. jemalloc) are very hard to beat with manual allocation strategies. We feel it is not worth picking this axe anymore.

For the specific case TVec/tvec! , it's a SmallVec of size 4, so it will use the stack to represent arrays of 4 elements or less, beyond that will switch to heap-allocated vector. It is used everywhere, for tensor shapes to operations inputs and outputs and much more.

**Comment by kali** [comment]
Very old (2019). Closing for housekeeping — see the examples/ dir; please open a fresh issue if you still need a streaming-audio exampl e.
